community/index/src/main/java/org/neo4j/index/gbptree/FreeListIdProvider.java

```diff
@@ -0,0 +1,351 @@
+ */
+package org.neo4j.index.gbptree;
+
+import java.io.IOException;
+import java.util.function.LongConsumer;
+
+import org.neo4j.io.pagecache.PageCursor;
+import org.neo4j.io.pagecache.PagedFile;
+
+import static org.neo4j.index.gbptree.PageCursorUtil.checkOutOfBounds;
+import static org.neo4j.index.gbptree.PageCursorUtil.goTo;
+
+class FreeListIdProvider implements IdProvider
+{
+    interface Monitor
+    {
+        /**
+         * Called when a page id was acquired for storing released ids into.
+         *
+         * @param freelistPageId page id of the acquired page.
+         */
+        default void acquiredFreelistPageId( long freelistPageId )
+        {   // Empty by default
+        }
+
+        /**
+         * Called when a free-list page was released due to all its ids being acquired.
+         * A released free-list page ends up in the free-list itself.
+         *
+         * @param freelistPageId page if of the released page.
+         */
+        default void releasedFreelistPageId( long freelistPageId )
+        {   // Empty by default
+        }
+    }
+
+    static final Monitor NO_MONITOR = new Monitor()
+    {   // Empty
+    };
+
+    private final PagedFile pagedFile;
+
+    /**
+     * {@link FreelistNode} governs physical layout of a free-list.
+     */
+    private final FreelistNode freelistNode;
+
+    /**
+     * There's one free-list which both stable and unstable state (the state pages A/B) shares.
+     * Each free list page links to a potential next free-list page, by using the last entry containing
+     * page id to the next.
+     * <p>
+     * Each entry in the the free list consist of a page id and the generation in which it was freed.
+     * <p>
+     * Read pointer cannot go beyond entries belonging to stable generation.
+     * About the free-list id/offset variables below:
+     * <pre>
+     * Every cell in picture contains generation, page id is omitted for briefness.
+     * StableGen   = 1
+     * UnstableGen = 2
+     *
+     *        {@link #readPos}                         {@link #writePos}
+     *        v                               v
+     *  ┌───┬───┬───┬───┬───┬───┐   ┌───┬───┬───┬───┬───┬───┐
+     *  │ 1 │ 1 │ 1 │ 2 │ 2 │ 2 │-->│ 2 │ 2 │   │   │   │   │
+     *  └───┴───┴───┴───┴───┴───┘   └───┴───┴───┴───┴───┴───┘
+     *  ^                           ^
+     *  {@link #readPageId}                  {@link #writePageId}
+     * </pre>
+     */
+    private volatile long writePageId;
+    private volatile long readPageId;
+    private volatile int writePos;
+    private volatile int readPos;
+
+    /**
+     * Last allocated page id, used for allocating new ids as more data gets inserted into the tree.
+     */
+    private volatile long lastId;
+
+    /**
+     * For monitoring internal free-list activity.
+     */
+    private final Monitor monitor;
+
+    FreeListIdProvider( PagedFile pagedFile, int pageSize, long lastId, Monitor monitor )
```

Comment:
ok i see i missed the branch with initializeAfterCreation call, its fine then to keep it like it is now